consensus/polybft/consensus_runtime.go

```diff
@@ -855,94 +835,239 @@ func (c *consensusRuntime) getCommitmentToRegister(epoch *epochMetadata,
 	}, nil
 }
 
-func validateVote(vote *MessageSignature, epoch *epochMetadata) error {
-	// get senders address
-	senderAddress := types.StringToAddress(string(vote.From))
-	if !epoch.Validators.ContainsAddress(senderAddress) {
-		return fmt.Errorf(
-			"message is received from sender %s, which is not in current validator set",
-			vote.From,
-		)
+func (c *consensusRuntime) IsValidBlock(proposal []byte) bool {
+	if err := c.fsm.Validate(proposal); err != nil {
+		c.logger.Error("failed validate proposal", "error", err)
+
+		return false
 	}
 
-	return nil
+	return true
+}
+
+func (c *consensusRuntime) IsValidSender(msg *proto.Message) bool {
+	err := c.fsm.ValidateSender(msg)
+	if err != nil {
+		c.logger.Error("invalid sender", "err", err)
+
+		return false
+	}
+
+	return true
+}
+
+func (c *consensusRuntime) IsProposer(id []byte, height, round uint64) bool {
+	nextProposer := c.fsm.validators.CalcProposer(round)
+
+	return bytes.Equal(id, nextProposer[:])
+}
+
+func (c *consensusRuntime) IsValidProposalHash(proposal, hash []byte) bool {
+	newBlock := types.Block{}
+	if err := newBlock.UnmarshalRLP(proposal); err != nil {
+		c.logger.Error("unable to unmarshal proposal", "err", err)
+
+		return false
+	}
+
+	return bytes.Equal(newBlock.Header.Hash.Bytes(), hash)
+}
+
+func (c *consensusRuntime) IsValidCommittedSeal(proposalHash []byte, committedSeal *messages.CommittedSeal) bool {
+	err := c.fsm.ValidateCommit(committedSeal.Signer, committedSeal.Signature, proposalHash)
+	if err != nil {
+		c.logger.Info("Invalid committed seal", "err", err)
+
+		return false
+	}
+
+	return true
 }
 
 func (c *consensusRuntime) BuildProposal(blockNumber uint64) []byte {
-	panic("not implemented")
+	lastBuiltBlock, _ := c.getLastBuiltBlockAndEpoch()
+
+	if lastBuiltBlock.Number+1 != blockNumber {
+		c.logger.Error("unable to build block, due to lack of parent block",
+			"last", lastBuiltBlock.Number, "num", blockNumber)
+
+		return nil
+	}
+
+	proposal, err := c.fsm.BuildProposal()
+	if err != nil {
+		c.logger.Info("Unable to create porposal", "blockNumber", blockNumber, "err", err)
+
+		return nil
```

Comment:
We'll return nil istead of proposal. Is it correct? 
Do we rely on nil value?

Reply:
The same thing is done for edge, to we followed the pattern.